Fix man.el shell injection vulnerability
Backported from emacs-30 branch
https://bugs.gentoo.org/950192
https://debbugs.gnu.org/66390

commit 820f0793f0b46448928905552726c1f1b999062f
Author: Xi Lu <lx@shellcodes.org>
Date:   Tue Oct 10 22:20:05 2023 +0800

    Fix man.el shell injection vulnerability

--- emacs-28.2/lisp/man.el
+++ emacs-28.2/lisp/man.el
@@ -688,7 +688,11 @@
       (setq name (match-string 2 ref)
 	    section (match-string 1 ref))))
     (if (string= name "")
-	ref				; Return the reference as is
+	;; see Bug#66390
+	(mapconcat #'identity
+		   (mapcar #'shell-quote-argument
+			   (split-string ref "\\s-+"))
+		   " ")			; Return the reference as is
       (if Man-downcase-section-letters-flag
 	  (setq section (downcase section)))
       (while slist
--- emacs-28.2/test/lisp/man-tests.el
+++ emacs-28.2/test/lisp/man-tests.el
@@ -161,6 +161,18 @@
           (let ((button (button-at (match-beginning 0))))
             (should (and button (eq 'Man-xref-header-file (button-type button))))))))))
 
+(ert-deftest man-tests-Man-translate-references ()
+  (should (equal (Man-translate-references "basename")
+                 "basename"))
+  (should (equal (Man-translate-references "basename(3)")
+                 "3 basename"))
+  (should (equal (Man-translate-references "basename(3v)")
+                 "3v basename"))
+  (should (equal (Man-translate-references ";id")
+                 "\\;id"))
+  (should (equal (Man-translate-references "-k basename")
+                 "-k basename")))
+
 (provide 'man-tests)
 
 ;;; man-tests.el ends here
